[DOC] Rewrite the TSDB CLI help; edit text in log messages, code comments, etc.

Sharon-iguazio opened the pull request.

Sharon-iguazio commented:
This PR mostly includes text changes and some related Cobra CLI-help code changes, but it also includes the following code changes:

- **pkg/tsdbctl/create.go** &mdash; add a successful table creation log message (Ln 163).

- **pkg/tsdbctl/delete.go**  &mdash;  rename the `delete` function's parameter from `ic` to `dc` (starting at Ln 88).
    > **Note:** I think it would have been better to name the commander function's parameter in all files in the same way, which isn't specific to the command name, such as `cmdr` (as I think `cmd` is already taken).

- **pkg/tsdbctl/query.go**  &mdash; 
  - Use the `DefaultOutputFormat` const (see the **pkg/formatter/type.go** changes) as the default value of the `-o|--output` flag (Ln 93).
  - Add mutually-exclusive flags validation for `-l|--last` and `-b|--before`/`'-e|--end` (Ln 177).

- **pkg/tsdbctl/time.go**  &mdash;
  - Remove the `put` alias.
  - Set the command-constructor `Hidden` parameter to `true` instead of adding `cmd.Hidden = true` at the end (Ln 43).

- **pkg/tsdbctl/tsdbctl.go**  &mdash;  add a `defaultVerbosityLevel` const (Ln 40) and use it as the default value of the `-v|--verbose` flag instead of using `"debug"` hardcoded (Ln 70).

- **pkg/formatter/type.go**  &mdash;  add a `DefaultOutputFormat` const.

- **pkg/tsdb/testdata/** &mdash; REMOVED.  The directory contained only an old version of the **v3io.yaml.template** TSDB CLI configuration file, which has been replaced with **examples/v3io.yaml.template**.
   > **Note:** I think GitHub commits apply only to files and not to directories, so the empty directory (after merging the PR) would need to be deleted manually from the upstream origin and local and fork branches.

Sharon-iguazio commented:
@imakhlin  Please review. (It seems I don't have GitHub permissions to assign you as the reviewer.)

imakhlin commented on examples/nuclio/query/query_example.go:
```diff
@@ -46,30 +47,33 @@ func Handler(context *nuclio.Context, event nuclio.Event) (interface{}, error) {
 	f, err := formatter.NewFormatter("json", nil)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to start json formatter")
+		return nil, errors.Wrap(err, "Failed to start the JSON formatter.")
```
Note, we usually tend to start error messages with lower case because of possibility of chaining multiple error messages into one string

Sharon-iguazio replied:
I changed all messages that seem to be standalone sentences to complete sentences that begin with a capital letter and usually end in a period. I think the overwhelming majority, if not all, of these messages are not concatenated into one sentence; (it's OK if they're concatenated into one message with multiple sentences). It looks back when you have a log message that's an actual sentence but begins with a lower-case letter.

Sharon-iguazio replied:
UPDATE: As discussed f2f, Eran D. says that log messages should be capitalized and while Go has a convention of beginning error messages with a lower-case letter, our convention is to capitalize error messages as well ==> my changes are consistent with our internal conventions. I made the changes in all (or almost all) non-test files.

imakhlin commented on pkg/config/config.go:
```diff
@@ -51,58 +51,68 @@ func Error() error {
-	RepotInterval      int    `json:"reportInterval"` // interval between consequence reports (in Seconds)
+	// Report on shutdown (Boolean)
+	ReportOnShutdown bool `json:"reportOnShutdown,omitempty"`
+	// Output destination - "stdout", "stderr", "syslog", etc.
```
please remove `, "syslog", etc.` as unsupported

Sharon-iguazio replied:
Done (although "syslog, etc." was in the original comment :-)).

imakhlin commented on pkg/tsdbctl/create.go:
```diff
@@ -62,7 +62,12 @@ func newCreateCommandeer(rootCommandeer *RootCommandeer) *createCommandeer {
-		Short: "create a new TSDB in the specifies path",
+		Short: "Create a new TSDB instance",
+		Long:  `Create a new TSDB instance (table) according to the provided configuration.`,
+		Example: `- tsdbctl create -s 192.168.1.100:8081 -u myuser -p mypassword -c mycontainer -t my_tsdb --rate 1/s
```
Change `--rate` to `--ingestion-rate`

Sharon-iguazio replied:
I saw the change, thanks. The help text for this flag in my PR is actually longer than what you used (that's why I wrote "..." at the end of the quoted text in my comment), but I'll handle this when merging the PR #141 changes from `development` to my PR branch.

imakhlin commented on pkg/tsdbctl/query.go:
```diff
@@ -66,15 +86,22 @@ func newQueryCommandeer(rootCommandeer *RootCommandeer) *queryCommandeer {
+	cmd.Flags().StringVarP(&commandeer.last, "last", "l", "",
+		"Return data for the specified time period before the\ncurrent time, of the format \"[0-9]+[mhd]\" (where\n'm' = minutes, 'h' = hours, and 'd' = days>). When setting\nthis flag, don't set the -b|--begin or -e|--end flags.\nExamples: \"1h\"; \"15m\"; \"30d\" to return data for the last\n1 hour, 15 minutes, or 30 days.")
+	cmd.Flags().StringVarP(&commandeer.windows, "windows", "w", "",
+		"Overlapping windows of time to which to apply the\naggregation functions (if set - see -a|--aggregates),\nstarting from the query's end time, as a comma separated\nlist of integer values (\"[0-9]+\"). The duration of each\nwindow is calculated by multiplying the flag value with\nthe aggregation interval (see -i|--aggregation-interval).\nExample: -w \"1,2\" with -i \"2h\" defines overlapping\na ...
```
How "1,2" became "2,4" ?
> Example: -w \"1,2\" with -i \"2h\" defines overlapping\naggegration windows of 2 hours and 4 hours.

Sharon-iguazio replied:
Because as I understood from the README (and verified with you f2f), and as I explain in the help text for the `-w|--windows` flag, the duration of each window is calculated by multiplying the value specified with the flag with the aggregation interval that can be specified with the `-i|--aggregation-interval` flag — so window values of `1` and `2` with an aggregation interval of `2h` should yield window durations of 2h and 4h. (The README gives an example of windows `1,2` with the default aggregation interval of `1h`, but I intentionally didn't want to use an interval of `1`, to clearly demonstrate the multiplication.) Did I misunderstand the behavior?

imakhlin replied:
incorrect. if step is smaller than a window - multiple rows will appear in the result-set for that window

Sharon-iguazio replied:
As we agreed on Slack, please provide an alternative explanation of the flag values (that we can use externally) and a relevant example with a step (aggregation interval) that's not 1h. Thanks.

Sharon-iguazio replied:
UPDATE: As we agreed today, the overlapping-windows documentation in the PR CLI help is correct.

imakhlin commented on pkg/tsdbctl/query.go:
```diff
@@ -152,12 +184,12 @@ func (qc *queryCommandeer) query() error {
 	f, err := formatter.NewFormatter(qc.output, nil)
 	if err != nil {
-		return errors.Wrap(err, "failed to start formatter "+qc.output)
+		return errors.Wrap(err, "Failed to start formatter '"+qc.output+"'.")
```
use `errors.Wrapf(err, "msg", params)`

Sharon-iguazio replied:
I don't understand exactly what you want me to change, and as this request is not related to the changes that I made (the code already used `Wrap`, I just edited the string), please make the changes that you want in a separate commit after we merge this PR.

imakhlin replied:
np

Sharon-iguazio replied:
So, I ended up making the change after I merged the `development` branch :-) — see commit 6e3a4dd.

imakhlin commented on pkg/tsdbctl/tsdbctl.go:
```diff
@@ -57,23 +58,34 @@ func NewRootCommandeer() *RootCommandeer {
+	cmd.PersistentFlags().StringVarP(&commandeer.v3ioPath, "server", "s", defaultV3ioServer,
+		"Web-gateway (web-APIs) service endpoint of an instance of\nthe Iguazio Continuous Data Platfrom, of the format\n\"<IP address>:<port number=8081>\". Examples: \"localhost:8081\"\n(when running on the target platform); \"192.168.1.100:8081\".")
+	cmd.PersistentFlags().StringVarP(&commandeer.cfgFilePath, "config", "g", "",
+		"Path to a YAML TSDB configuration file. When this flag isn't\nset, the CLI checks for a "+config.DefaultConfigurationFileName+" configuration file in the\ncurrent directory. CLI flags override file confiugrations.\nExample: \"~/cfg/my_v3io_tsdb_cfg.yaml\".")
```
I think word `YAML` is useless in this sentence

Sharon-iguazio replied:
I think it's helpful because it clarifies in which format we expect the configuration file to be.

Sharon-iguazio commented on pkg/tsdbctl/create.go:
```diff
@@ -91,16 +100,16 @@ func (cc *createCommandeer) create() error {
 
 	if cc.sampleRate == "" {
-		return errors.New(`sample rate not provided! Please provide sample rate with --rate flag in the format of [0-9]+/[hms]. Example: 12/m`)
+		return errors.New(`Sample rate not provided. Use the --rate flag to provide a sample rate in the format of "[0-9]+/[mhd]". For example, "12/m".`)
```
Done — see commit bf933d2. I also renamed the `sampleRate` struct field and function parameter in **create.go** to `samplesIngestionRate` and I edited related error messages.

Sharon-iguazio commented on pkg/tsdbctl/create.go:
```diff
@@ -198,7 +211,7 @@ func (cc *createCommandeer) calculatePartitionAndChunkInterval(rateInHours int)
 
 func rateToHours(sampleRate string) (int, error) {
-	parsingError := errors.New(`not a valid rate. Accepted pattern: [0-9]+/[hms]. Example: 12/m`)
+	parsingError := errors.New(`Invalid rate. The sample rate must be of the format "[0-9]+/[mhd]". For example, "12/m".`)
```
Done — see commit bf933d2.

Sharon-iguazio commented on pkg/tsdbctl/info.go:
```diff
@@ -41,16 +41,21 @@ func newInfoCommandeer(rootCommandeer *RootCommandeer) *infoCommandeer {
 	cmd := &cobra.Command{
 		Use:   "info",
-		Short: "return TSDB config",
+		Short: "Return information about a TSDB instance",
```
Done — commit e1053ad — but I used "Display" and not "Print"; I also edited the flag descriptions accordingly (I also was didn't really like "Return", but I assumed you liked it because it was used in the existing help text so I decided to leave it.)
Note that the convention of the TSDB CLI help text is to phrase the description as what the command should do and now what it does; this was the existing convention and I didn't change it. So it should be "Print"/"Display" and not "Prints"/"Displays".

Sharon-iguazio commented:
Please see new commit 535466e — in **create.go**, I replaced the "rollup interval" terminology in local code elements with "aggregation granularity" terminology + I edited the related error message.

imakhlin commented on pkg/tsdbctl/create.go:
```diff
@@ -44,15 +44,15 @@ const (
+	rollupInterval          string
+	shardingBuckets         int
+	sampleRetention         int
+	samplesIngestionRate    string
```
try to make less changes to the code to avoid troubles with merge conflicts

Sharon-iguazio commented:
@imakhlin  Please see the merge changes in commits 1cecaa5 and 4e230c7 (which mostly merges changes previously done in this PR in **pkg/tsdbctl/create.go** to the new **pkg/tsdb/schema/schema.go** file into which you moved the relevant code).

imakhlin commented on README.md:
```diff
@@ -132,18 +132,20 @@ parameters: the configuration structure, v3io data container object and logger o
-or can be loaded from a local buffer using `config.GetOrLoadFromData(data []byte)`. You can see details on the configuration
-options in [config](internal/pkg/config/config.go), a minimal configuration looks like: 
+or can be loaded from a local buffer using `config.GetOrLoadFromData(data []byte)`.
+You can see details on the configuration options in [config](internal/pkg/config/config.go).
```
Path: `$GOPATH/src/github.com/v3io/v3io-tsdb/pkg/config/config.go`

Sharon-iguazio replied:
The config link target wasn't edited by me; I also don't like the text, but I didn't do a full review of the README file yet. But as you've pointed it out, I've now changed the sentence to the following:
> You can see details on the configuration options in the V3IO TSDB [**config.go**](pkg/config/config.go) source file.
Note that I removed `internal/` from the link target and left it as `pkg/config/config.go`. You can't use `$GOPATH/...` in a Mardkown link target. Using a relative file path is the best practice for such cases and works well. When reading the README file in the GH web interface, the link redirects to the relative target config file for the repo branch that you're viewing, which is what you want.

imakhlin commented on pkg/formatter/type.go:
```diff
@@ -10,12 +10,14 @@ import (
 	"time"
 )
 
+const DefaultOutputFormat = "text"
```
should be private, i.e. starts with lower-case

imakhlin replied:
ok, I see that you're using this constant in query.go
Why didn't you define constants for all formats? If you decide to improve something - do it all the way.

Note, the test in CLI help contains quotes where it refers to the string values. For example:
```
 -o, --output string                 Output format in which to display the query results -
                                      "text" | "csv" | "json". (default "text")
```
There is no need to use quotes in CLI, so why did you do that?

Sharon-iguazio replied:
1. I don't think there's a need for consts for all supported types. I think we should, however, have a const to define what's the default value, as we have for other default values, which is what I added. (Also, it's not my job to improve the CLI code and it's OK to improve it in stages ... .)
2. While the quotes around string CLI arguments aren't required (unless the values have spaces), they're not wrong, and I prefer to add them in the CLI help for clarity, especially seeing as there's no text formatting in the CLI. This is also consistent with the Cobra's behavior, as it adds quotes around the default values of string flags (e.g., `(default "text")` in this case was added by Cobra). I did omit the strings from the supported aggregate functions list (after going back and forth) because there the values are names of functions that the final flag string can optionally contain (+ it's a long list and adding the quotes seems to cumbersome).

imakhlin commented on pkg/tsdbctl/query.go:
```diff
@@ -66,15 +86,22 @@ func newQueryCommandeer(rootCommandeer *RootCommandeer) *queryCommandeer {
+		"End (maximum) time for the query, as a string containing an\nRFC3339 time string, a Unix timestamp in milliseconds, or\na relative time of the format \"now\" or \"now-[0-9]+[mhd]\"\n(where 'm' = minutes, 'h' = hours, and 'd' = \"days\").\nExamples: \"2018-09-26T14:10:20Z\"; \"1537971006000\";\n\"now-3h\"; \"now-7d\". (default \"now\")")
+	cmd.Flags().StringVarP(&commandeer.from, "begin", "b", "",
+		"Start (minimum) time for the query, as a string containing\nan RFC3339 time, a Unix timestamp in milliseconds, a\nrelative time of the format \"now\" or \"now-[0-9]+[mhd]\"\n(where 'm' = minutes, 'h' = hours, and 'd' = \"days\"), or 0\nfor the earliest time. Examples: \"2016-01-02T15:34:26Z\";\n\"1451748866\"; \"now-90m\"; \"0\". (default = <end time> - 1h)")
+	cmd.Flags().StringVarP(&commandeer.output, "output", "o", formatter.DefaultOutputFormat,
```
Why did you change the "default" value?

Sharon-iguazio replied:
I'm now sure which change you're referring to. The original code didn't document the default values for the `query` end and start time flags. I first added documentation of the default value for the `-e|--end` flag (commit 15fd9b4) and then for the `'-b|--begin` flag (commit e64d18e).

imakhlin commented on pkg/tsdb/schema/schema.go:
```diff
@@ -15,53 +15,53 @@ const (
-	if err := validateAggregatorGranularity(aggregatorGranularity); err != nil {
-		return nil, errors.Wrap(err, "failed to parse aggregator granularity")
+	if err := validateAggregatesGranularity(aggregationGranularity); err != nil {
+		return nil, errors.Wrapf(err, "Failed to parse aggregation granularity '%v'.", aggregationGranularity)
```
use `%s` for `string` arguments

Sharon-iguazio replied:
Done — see commit 7830f78.

imakhlin commented on pkg/tsdbctl/tsdbctl.go:
```diff
@@ -53,23 +53,34 @@ func NewRootCommandeer() *RootCommandeer {
-	cmd.PersistentFlags().StringVarP(&commandeer.password, "password", "p", "", "password")
-
-	// add children
+	cmd.PersistentFlags().StringVarP(&commandeer.verbose,
```
`verbose` was changed to `logLevel`

Sharon-iguazio replied:
MERGED `development` with my previous PR changes (including using `DefaultLoggingLevel`, which I had defined in *config.go* — info in comment https://github.com/v3io/v3io-tsdb/pull/148#issuecomment-426176080). I edited the flag's help text in *tsdbctl.go* and the related configuration-file `logLevel` key code coment *config.go* (including to remove the referene to`info` being the default logging level, as we discussed on Slack). See commit 2d22ad7.

Sharon-iguazio commented on pkg/tsdbctl/add.go:
```diff
@@ -55,13 +56,36 @@ func newAddCommandeer(rootCommandeer *RootCommandeer) *addCommandeer {
+		Example: `The examples assume that the endpoint of the web-gateway service, the login credentails, and
+the name of the data container are configured in the default configuration file (` + config.DefaultConfigurationFileName + `)
+instead of using the -s|--server, -u|--username, -p|--password, and -c|--container flags.
+- tsdbctl -t mytsdb add temperature -d 28 -m now-2h
```
Already done — I moved command names and arguments before flags in all the examples; see commit 207d08b (in which I also added a `query` overlapping-windows examples and edited the `-w|--windows` flag doc) and commit a4c2637.

Sharon-iguazio commented:
**DO NOT MERGE** yet — there's incorrect query behavior. I'm debugging it.

Sharon-iguazio commented on pkg/tsdbctl/check.go:
```diff
@@ -43,23 +44,32 @@ func newCheckCommandeer(rootCommandeer *RootCommandeer) *checkCommandeer {
+		Long:   `Get information about a TSDB metric item.`,
+		Example: `The examples assume that the endpoint of the web-gateway service, the login credentails, and
+the name of the data container are configured in the default configuration file (` + config.DefaultConfigurationFileName + `)
+- tsdbctl -t mytsdb 1537920000/cpu
```
@gshatz I added a `tsdbctl check 1538265600/memo_1.c6b54e7ce82c2c11 -t my_tsdb -a _v12` example based on your example. I also edited the existing no-attributes example to change the item path to `1537920000/cpu_2.40480e29c7d6f962` (taken from my `_tsdbtest2` table on Dev2) and add the `check` command that was missing by mistake.
In the description of the `-a|--attrs` flag, I added the examples `_v35` and `"_v22,_v23"` (which I successfully tested with items in my test tables), and I edited the text to refer specifically to _blob_ attributes. See commit a6feecc.
@imakhlin Please review these changes as well.

Sharon-iguazio replied:
@gshatz Following our Slack conversation, I changed the order  of the example and edited the no `-a` example to also use `-a`  but with multiple attributes — `tsdbctl check 1537920000/cpu_2.40480e29c7d6f962 -t mytsdb -a "_v30,_v31"`. I also made the `-a|--attrs` flag required (tested successfully). See commit 5d8e989. @imakhlin  FYI.

Sharon-iguazio commented:
@imakhlin  @gshatz I finished responding to all your comments and implementing relevant changes + I merged all `development` branch changes. The PR currently passes all the DevOps tests + I had successfully tested all of the CLI options documented in the help (and more). -->Please review my latest commits (see details in the PR comments), resolve all the open PR conversations, and approve and merge the PR, unless you have more comments. Thanks.

imakhlin commented:
LGTM, Please resolve conflicts and go ahead

Sharon-iguazio commented:
@imakhlin I renamed `verbose` in the *examples/v3io.yaml.template* template config file to `logLevel` (to match the code) and I removed the default "debug" level comment; as we discussed on Slack, this is the default only when using the `-v|--log-level` CLI flag without specifying the debug level. See commit 8d319a9.
